### PR TITLE
[wip] Seq2SeqTrainer

### PR DESCRIPTION
**Pull request by patil-suraj**
cc @sshleifer

### CODE REVIEW

File: examples/seq2seq/seq2seq_trainer.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            warnings.warn(
+                "The `_training_step` method is deprecated and won't be called in a future version, define `training_step` in your subclass.",
+                FutureWarning,
+            )
+            return self._training_step(model, inputs, self.optimizer)
+
+        model.train()
+        inputs = self._prepare_inputs(inputs, model)
+
+        if self.args.fp16 and _use_native_amp:
```
Comment (sshleifer) [inline examples/seq2seq/seq2seq_trainer.py:57]: this not mutually exclusive with label smoothing.

Comment (sshleifer) [reply examples/seq2seq/seq2seq_trainer.py:57]: will apex fp16 work automagically?

File: examples/seq2seq/seq2seq_trainer.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            with autocast():
+                outputs = model(**inputs)
+                loss = outputs[0]
+        else:
+            outputs = model(**inputs)
+            # We don't use .loss here since the model may return tuples instead of ModelOutput.
+            loss = outputs[0]
+
+        if self.args.past_index >= 0:
+            self._past = outputs[self.args.past_index]
```
Comment (sshleifer) [inline examples/seq2seq/seq2seq_trainer.py:67]: Also make sure passing `use_cache=False` to forward.
where is this used? We dont need it for forward pass.

File: examples/seq2seq/seq2seq_trainer.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        if self.args.n_gpu > 1:
+            loss = loss.mean()  # mean() to average on multi-gpu parallel training
+
+        if self.args.gradient_accumulation_steps > 1:
+            loss = loss / self.args.gradient_accumulation_steps
+        
+        # assuming label_smoothing is in args
+        if self.args.label_smoothing == 0:
+            outputs = model(**inputs)
+            loss = outputs[0]  # model outputs are always tuple in transformers (see doc)
```
Comment (sshleifer) [inline examples/seq2seq/seq2seq_trainer.py:78]: I wouldn't trust the model to do this. It might not ignore pad_token_id.

File: examples/seq2seq/seq2seq_trainer.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        else:
+            labels = inputs.pop("labels")
+            labels[labels == -100] = model.config.pad_token_id
```
Comment (sshleifer) [inline examples/seq2seq/seq2seq_trainer.py:81]: did somebody put 100 in labels?

**Review by sshleifer — COMMENTED** [review]
cool start, tag me know when you want more comments/help!

File: examples/seq2seq/seq2seq_trainer.py (hunk 1 above)
Comment (patil-suraj) [reply examples/seq2seq/seq2seq_trainer.py:57]: > will apex fp16 work automagically?

It should, right ? The `Triner` takes care of it. I've tried it and worked in my case, but will test it to make sure.

File: examples/seq2seq/seq2seq_trainer.py (hunk 3 above)
Comment (patil-suraj) [reply examples/seq2seq/seq2seq_trainer.py:78]: I'm thinking of adding `Seq2SeqDataCollator` which will replace `pad_tokens` with -100 (ignore index), shift `labels` depending on the mode , prepare right `decoder_input_ids` etc. Let me know if this is right approach.

File: examples/seq2seq/seq2seq_trainer.py (hunk 4 above)
Comment (patil-suraj) [reply examples/seq2seq/seq2seq_trainer.py:81]: collator
